src/lib.rs

```diff
@@ -1410,6 +1480,356 @@ where
+            serde_json::from_value(rsp);
+        val.map_err(|err| RequestTokenError::Parse(err, body))
+    })
+}
+
+///
+/// Maps from an HTTP response to a DeviceCodeAction, as some HTTP responses
+/// indicate that further request attempts should be made.
+///
+fn device_token_action(
+    http_response: HttpResponse,
+) -> Result<HttpResponse, DeviceCodeAction<HttpResponse>> {
+    if http_response.status_code != StatusCode::OK {
+        let reason = http_response.body.as_slice();
+        if reason.is_empty() {
+            Err(DeviceCodeAction::NoFurtherRequests(
+                http_response,
+            ))
+        } else {
+            match serde_json::from_slice::<DeviceCodeErrorResponse>(reason) {
+                Ok(error) => Err(error.to_action(http_response)),
+                Err(_) => Err(DeviceCodeAction::NoFurtherRequests(
+                    http_response,
+                )),
+            }
+        }
+    } else {
+        Ok(http_response)
+    }
+}
+
+///
+/// The request for an device access token from the authorization server.
+///
+/// See https://tools.ietf.org/html/rfc8628#section-3.4.
+///
+#[derive(Clone, Debug)]
+pub struct DeviceAccessTokenRequest<'a, TE, TR, TT>
+where
+    TE: ErrorResponse,
+    TR: TokenResponse<TT>,
+    TT: TokenType,
+{
+    auth_type: &'a AuthType,
+    client_id: &'a ClientId,
+    client_secret: Option<&'a ClientSecret>,
+    extra_params: Vec<(Cow<'a, str>, Cow<'a, str>)>,
+    token_url: Option<&'a TokenUrl>,
+    device_authorization_details: Option<&'a DeviceAuthorizationDetails>,
+    _phantom: PhantomData<(TE, TR, TT)>,
+}
+
+impl<'a, TE, TR, TT> DeviceAccessTokenRequest<'a, TE, TR, TT>
+where
+    TE: ErrorResponse + 'static,
+    TR: TokenResponse<TT>,
+    TT: TokenType,
+{
+    ///
+    /// Appends an extra param to the token request.
+    ///
+    /// This method allows extensions to be used without direct support from
+    /// this crate. If `name` conflicts with a parameter managed by this crate, the
+    /// behavior is undefined. In particular, do not set parameters defined by
+    /// [RFC 6749](https://tools.ietf.org/html/rfc6749) or
+    /// [RFC 7636](https://tools.ietf.org/html/rfc7636).
+    ///
+    /// # Security Warning
+    ///
+    /// Callers should follow the security recommendations for any OAuth2 extensions used with
+    /// this function, which are beyond the scope of
+    /// [RFC 6749](https://tools.ietf.org/html/rfc6749).
+    ///
+    pub fn add_extra_param<N, V>(mut self, name: N, value: V) -> Self
+    where
+        N: Into<Cow<'a, str>>,
+        V: Into<Cow<'a, str>>,
+    {
+        self.extra_params.push((name.into(), value.into()));
+        self
+    }
+
+    ///
+    /// Synchronously polls the authorization server for a response.
+    ///
+    pub fn request<F, RE>(self, http_client: F) -> Result<TR, RequestTokenError<RE, TE>>
+    where
+        F: FnOnce(HttpRequest) -> Result<HttpResponse, RE> + Copy,
```

Comment:
Have realised that this wants to be
```
F: Fn(HttpRequest) -> Result<HttpResponse, RE>,
```
rather than
```
F: FnOnce(HttpRequest) -> Result<HttpResponse, RE> + Copy,
```
, will fix